depthai_ros_driver/include/depthai_ros_driver/depthai_common.hpp

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef DEPTHAI_ROS_DRIVER__DEPTHAI_COMMON_HPP
+#define DEPTHAI_ROS_DRIVER__DEPTHAI_COMMON_HPP
+
+
+// Common ros and ros2 api
+#include <cv_bridge/cv_bridge.h>
+
+// relevant 3rd party includes
+#include <depthai/device.hpp>
+#include <depthai/host_data_packet.hpp>
+#include <depthai/nnet/nnet_packet.hpp>
+#include <depthai/pipeline/cnn_host_pipeline.hpp>
+
+// std includes
+#include <algorithm>
+#include <array>
+#include <memory>
+#include <string>
+#include <variant>
+
+#if defined(USE_ROS2)
+  #include <std_msgs/msg/float32.hpp>
+  #include <sensor_msgs/msg/image.hpp>
+  #include <depthai_ros_msgs/srv/trigger_named.hpp>
+  #include <depthai_ros_msgs/msg/objects.hpp>
+  #include <depthai_ros_msgs/msg/auto_focus_ctrl.hpp>
+  #include <camera_info_manager/camera_info_manager.hpp>
+  #include <rclcpp/node.hpp>
+
+  #define ROS_LOGGER(...) RCUTILS_LOG_WARN_NAMED("depthai_common", __VA_ARGS__)
+
+using ImageMsg = sensor_msgs::msg::Image;
+using CompressedImageMsg = sensor_msgs::msg::CompressedImage;
+using ObjectsMsg = depthai_ros_msgs::msg::Objects;
+using ObjectMsg = depthai_ros_msgs::msg::Object;
+using CameraInfoMsg = sensor_msgs::msg::CameraInfo;
+using HeaderMsg = std_msgs::msg::Header;
+using ROSNodeHandle = std::shared_ptr<rclcpp::Node>;
+
+using ObjectPubPtr = rclcpp::Publisher<ObjectMsg>::SharedPtr;
+using ObjectsPubPtr = rclcpp::Publisher<ObjectsMsg>::SharedPtr;
+using ImagePubPtr = rclcpp::Publisher<ImageMsg>::SharedPtr;
+using ComImagePubPtr = rclcpp::Publisher<CompressedImageMsg>::SharedPtr;
+using StreamPub = std::variant<ObjectsPubPtr, ObjectPubPtr,
```

Comment:
Instead of using a variant, can we try using type erasure that I discussed before (not in this PR, in a separate PR targetting this branch).

I want to compare the core code difference.

Reply:
I'm not sure there is any benefit in directly using type erasure here. `std::variant` uses type erasure internally; effectively that's what it's for.

Reply:
variant uses a switch for type safety, but we can test that later. it's a small piece of code